flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java

```diff
@@ -697,13 +738,19 @@ protected static void initializeSubscribedPartitionsToStartOffsets(
 			int indexOfThisSubtask,
 			int numParallelSubtasks,
 			StartupMode startupMode,
+			Date specificStartupDate,
 			Map<KafkaTopicPartition, Long> specificStartupOffsets) {
 
 		for (int i = 0; i < kafkaTopicPartitions.size(); i++) {
 			if (i % numParallelSubtasks == indexOfThisSubtask) {
-				if (startupMode != StartupMode.SPECIFIC_OFFSETS) {
-					subscribedPartitionsToStartOffsets.put(kafkaTopicPartitions.get(i), startupMode.getStateSentinel());
-				} else {
+				if (startupMode == StartupMode.SPECIFIC_TIMESTAMP) {
+					if (specificStartupDate == null) {
+						throw new IllegalArgumentException(
+							"Startup mode for the consumer set to " + StartupMode.SPECIFIC_TIMESTAMP +
+								", but no specific timestamp were specified");
+					}
+					subscribedPartitionsToStartOffsets.put(kafkaTopicPartitions.get(i), specificStartupDate.getTime());
```

Comment:
This is the main problem:
following the original design pattern, it would be better to place a `KafkaTopicPartitionStateSentinel` here instead of eagerly converting the `Date` to a specific offset. We only convert the date to specific offsets when we're about to start consuming the partition (i.e. in `KafkaConsumer` thread).